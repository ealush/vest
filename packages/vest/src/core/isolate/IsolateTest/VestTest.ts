import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
