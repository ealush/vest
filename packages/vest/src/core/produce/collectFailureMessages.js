import isMatchingSeverityProfile from 'isMatchingSeverityProfile';
import useTestObjects from 'useTestObjects';

/**
 * @param {'warn'|'error'} severity   Filter by severity.
 * @param {Object} options
 * @param {String} [options.group]      Group name for error lookup.
 * @param {String} [options.fieldName]  Field name for error lookup.
 * @returns all messages for given criteria.
 */
const collectFailureMessages = (severity, { group, fieldName } = {}) => {
  const [testObjects] = useTestObjects();
  return testObjects.reduce((collector, testObject) => {
    if (group && testObject.groupName !== group) {
      return collector;
    }

    if (fieldName && testObject.fieldName !== fieldName) {
      return collector;
    }

    if (!testObject.failed) {
      return collector;
    }

    if (isMatchingSeverityProfile(severity, testObject)) {
      return collector;
    }

    collector[testObject.fieldName] = (
      collector[testObject.fieldName] || []
    ).concat(testObject.statement);

    return collector;
  }, {});
};

export default collectFailureMessages;
