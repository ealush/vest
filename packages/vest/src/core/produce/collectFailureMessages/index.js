import { SEVERITY_GROUP_WARN } from '../../test/lib/VestTest/constants';

/**
 * @param {Object} State            State Object reference
 * @param {'warn'|'error'} severity   Filter by severity.
 * @param {Object} options
 * @param {String} [options.group]      Group name for error lookup.
 * @param {String} [options.fieldName]  Field name for error lookup.
 * @returns all messages for given criteria.
 */
const collectFailureMessages = (state, severity, { group, fieldName } = {}) =>
  state.testObjects.reduce((collector, testObject) => {
    if (group && testObject.groupName !== group) {
      return collector;
    }

    if (fieldName && testObject.fieldName !== fieldName) {
      return collector;
    }

    if (!testObject.failed) {
      return collector;
    }

    if (
      (severity !== SEVERITY_GROUP_WARN && testObject.isWarning) ||
      (severity === SEVERITY_GROUP_WARN && !testObject.isWarning)
    ) {
      return collector;
    }

    collector[testObject.fieldName] = (
      collector[testObject.fieldName] || []
    ).concat(testObject.statement);

    return collector;
  }, {});

export default collectFailureMessages;
