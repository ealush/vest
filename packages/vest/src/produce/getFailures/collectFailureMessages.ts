import VestTest from 'VestTest';
import nonMatchingFieldName from 'nonMatchingFieldName';
import nonMatchingSeverityProfile from 'nonMatchingSeverityProfile';
import type { TSeverity } from 'vestTypes';

export default function collectFailureMessages(
  severity: TSeverity,
  testObjects: VestTest[],
  options: { group?: string; fieldName?: string } = {}
): Record<string, string[]> {
  const { group, fieldName } = options || {};
  const res = testObjects.reduce(
    (collector: Record<string, string[]>, testObject) => {
      if (noMatch(testObject, severity, group, fieldName)) {
        return collector;
      }

      if (!testObject.failed) {
        return collector;
      }

      collector[testObject.fieldName] = (
        collector[testObject.fieldName] || []
      ).concat(testObject.message || []);

      return collector;
    },
    { ...(fieldName && { [fieldName]: [] }) }
  );

  return res;
}

function noGroupMatch(
  testObject: VestTest,
  groupName?: void | string
): boolean {
  return !!(groupName && testObject.groupName !== groupName);
}

function noMatch(
  testObject: VestTest,
  severity: TSeverity,
  group: void | string,
  fieldName: void | string
): boolean {
  if (noGroupMatch(testObject, group)) {
    return true;
  }

  if (nonMatchingFieldName(testObject, fieldName)) {
    return true;
  }

  if (nonMatchingSeverityProfile(severity, testObject)) {
    return true;
  }

  return false;
}
