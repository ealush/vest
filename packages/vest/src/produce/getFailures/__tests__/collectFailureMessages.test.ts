import VestTest from 'VestTest';
import collectFailureMessages from 'collectFailureMessages';

describe('collectFailureMessages', () => {
  let testObjects: VestTest[] = [];

  it('Should return an object containing just the requested field', () => {
    const result = collectFailureMessages('errors', testObjects, {
      fieldName: 'field_1',
    });
    expect(Object.keys(result)).toEqual(['field_1']);
  });

  test('Result has an array of matching error messages', () => {
    const result = collectFailureMessages('errors', testObjects, {
      fieldName: 'field_1',
    });
    expect(result.field_1).toEqual([
      'field_1_failure message 2',
      'field_1_failure message 3',
    ]);
  });

  it('Should return filtered messages by the selected group', () => {
    const result = collectFailureMessages('errors', testObjects, {
      group: 'group1',
    });

    expect(result).toEqual({ field_2: ['field_2_failure message 3'] });
  });

  it('Should return an empty object when no options and no failures', () => {
    expect(
      collectFailureMessages('errors', [
        new VestTest('field_1', jest.fn(), { message: 'error_message' }),
      ])
    ).toEqual({});
  });

  it('Should return an object with an empty array when selected field has no errors', () => {
    expect(
      collectFailureMessages(
        'errors',
        [new VestTest('field_1', jest.fn(), { message: 'error_message' })],
        { fieldName: 'field_1' }
      )
    ).toEqual({ field_1: [] });
  });

  describe.each(['errors', 'warnings'])(
    'Snapshot tests. severity: %s',
    // @ts-ignore - it is unable to infer errors/warnings correctly
    (severity: 'errors' | 'warnings') => {
      describe('When no options passed', () => {
        it('should match snapshot', () => {
          expect(
            collectFailureMessages(severity, testObjects)
          ).toMatchSnapshot();
        });
      });

      describe('When specific field requested', () => {
        it('Should match snapshot', () => {
          expect(
            collectFailureMessages(severity, testObjects, {
              fieldName: 'field_1',
            })
          ).toMatchSnapshot();
          expect(
            collectFailureMessages(severity, testObjects, {
              fieldName: 'field_2',
            })
          ).toMatchSnapshot();
          expect(
            collectFailureMessages(severity, testObjects, {
              fieldName: 'field_3',
            })
          ).toMatchSnapshot();
        });
      });

      describe('When specific group requested', () => {
        it('Should match snapshot', () => {
          expect(
            collectFailureMessages(severity, testObjects, {
              group: 'group1',
            })
          ).toMatchSnapshot();
          expect(
            collectFailureMessages(severity, testObjects, {
              group: 'group1',
            })
          ).toMatchSnapshot();
          expect(
            collectFailureMessages(severity, testObjects, {
              group: 'group1',
            })
          ).toMatchSnapshot();
        });
      });

      describe('When a group-fieldName combo is requested', () => {
        it('Should match snapshot', () => {
          expect(
            collectFailureMessages(severity, testObjects, {
              group: 'group1',
              fieldName: 'field_2',
            })
          ).toMatchSnapshot();
        });
      });
    }
  );

  beforeEach(() => {
    testObjects = [
      new VestTest('field_1', jest.fn(), {
        message: 'field_1_failure message 1',
      }),
      (() => {
        const test = new VestTest('field_1', jest.fn(), {
          message: 'field_1_failure message 2',
        });
        test.fail();
        return test;
      })(),
      (() => {
        const test = new VestTest('field_1', jest.fn(), {
          message: 'field_1_failure message 3',
        });
        test.fail();
        return test;
      })(),
      new VestTest('field_1', jest.fn(), { groupName: 'group1' }),
      new VestTest('field_2', jest.fn(), {
        groupName: 'group1',
        message: 'field_2_failure message 1',
      }),
      (() => {
        const test = new VestTest('field_2', jest.fn(), {
          message: 'field_2_warning message 1',
        });
        test.warn();
        return test;
      })(),
      (() => {
        const test = new VestTest('field_2', jest.fn(), {
          message: 'field_2_warning message 2',
        });
        test.fail();
        test.warn();
        return test;
      })(),
      (() => {
        const test = new VestTest('field_2', jest.fn(), {
          groupName: 'group1',
          message: 'field_2_warning message 3',
        });
        test.fail();
        test.warn();
        return test;
      })(),
      (() => {
        const test = new VestTest('field_2', jest.fn(), {
          groupName: 'group1',
          message: 'field_2_failure message 3',
        });
        test.fail();
        return test;
      })(),
      new VestTest('field_3', jest.fn()),
    ];
  });
});
