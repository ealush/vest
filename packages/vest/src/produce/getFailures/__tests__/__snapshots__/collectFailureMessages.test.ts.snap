// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectFailureMessages Snapshot tests. severity: errors When a group-fieldName combo is requested Should match snapshot 1`] = `
Object {
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When no options passed should match snapshot 1`] = `
Object {
  "field_1": Array [
    "field_1_failure message 2",
    "field_1_failure message 3",
  ],
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific field requested Should match snapshot 1`] = `
Object {
  "field_1": Array [
    "field_1_failure message 2",
    "field_1_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific field requested Should match snapshot 2`] = `
Object {
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific field requested Should match snapshot 3`] = `
Object {
  "field_3": Array [],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific group requested Should match snapshot 1`] = `
Object {
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific group requested Should match snapshot 2`] = `
Object {
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: errors When specific group requested Should match snapshot 3`] = `
Object {
  "field_2": Array [
    "field_2_failure message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When a group-fieldName combo is requested Should match snapshot 1`] = `
Object {
  "field_2": Array [
    "field_2_warning message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When no options passed should match snapshot 1`] = `
Object {
  "field_2": Array [
    "field_2_warning message 2",
    "field_2_warning message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific field requested Should match snapshot 1`] = `
Object {
  "field_1": Array [],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific field requested Should match snapshot 2`] = `
Object {
  "field_2": Array [
    "field_2_warning message 2",
    "field_2_warning message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific field requested Should match snapshot 3`] = `
Object {
  "field_3": Array [],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific group requested Should match snapshot 1`] = `
Object {
  "field_2": Array [
    "field_2_warning message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific group requested Should match snapshot 2`] = `
Object {
  "field_2": Array [
    "field_2_warning message 3",
  ],
}
`;

exports[`collectFailureMessages Snapshot tests. severity: warnings When specific group requested Should match snapshot 3`] = `
Object {
  "field_2": Array [
    "field_2_warning message 3",
  ],
}
`;
